File: 프로그래머스/Lv.4/43236. 징검다리/징검다리.py
def is_passable(target_distance, rocks, n):
    removed = last = 0
    
    for rock in rocks:
        if rock - last < target_distance:
            removed += 1
            continue
        last = rock
        
    return removed <= n


def solution(distance, rocks, n):
    rocks.append(distance)
    rocks.sort()
    
    min_distance, max_distance = 1, distance + 1
    
    while min_distance + 1 < max_distance:
        target_distance = (min_distance + max_distance) // 2
        
        if is_passable(target_distance, rocks, n):
            min_distance = target_distance
        else:
            max_distance = target_distance
    
    return min_distance

File: 프로그래머스/Lv.4/43236. 징검다리/징검다리.java
import java.util.Arrays;

class Solution {
    
    public int solution(int distance, int[] rocks, int n) {
        int answer = 1;
        int minDistance = 1;
        int maxDistance = distance + 1;
        
        rocks = Arrays.copyOf(rocks, rocks.length + 1);
        rocks[rocks.length - 1] = distance;
        Arrays.sort(rocks);
        
        while (minDistance < maxDistance) {
            int targetDistance = (minDistance + maxDistance) / 2;
            
            if (canCrossBridge(targetDistance, rocks, n)) {
                answer = Math.max(answer, targetDistance);
                minDistance = targetDistance + 1;
            } else {
                maxDistance = targetDistance;
            }
        }
        
        return answer;
    }
    
    private boolean canCrossBridge(int targetDistance, int[] rocks, int n) {
        int removed = 0;
        int last = 0;
        
        for (int rock : rocks) {
            if (rock - last < targetDistance) {
                removed++;
                continue;
            }
            last = rock;
        }
        
        return removed <= n;
    }
}
